source/darwin/dispatch_queue_event_loop.c

```diff
@@ -0,0 +1,698 @@
+    struct aws_ref_count ref_count;
+};
+
+/**
+ * The data structure used to track the dispatch queue execution iteration (block). Each entry associated to an
+ * iteration scheduled on Apple Dispatch Queue.
+ */
+struct scheduled_service_entry {
+    struct aws_allocator *allocator;
+    uint64_t timestamp;
+    struct aws_priority_queue_node priority_queue_node;
+    struct dispatch_loop_context *dispatch_queue_context;
+};
+
+/** Help functions to track context ref-count */
+
+static void *s_acquire_dispatch_loop_context(struct dispatch_loop_context *context) {
+    return aws_ref_count_acquire(&context->ref_count);
+}
+
+static size_t s_release_dispatch_loop_context(struct dispatch_loop_context *context) {
+    return aws_ref_count_release(&context->ref_count);
+}
+
+/** Help functions to lock status */
+static int s_rlock_dispatch_loop_context(struct dispatch_loop_context *context) {
+    return aws_rw_lock_rlock(&context->lock);
+}
+
+static int s_runlock_dispatch_loop_context(struct dispatch_loop_context *context) {
+    return aws_rw_lock_runlock(&context->lock);
+}
+
+static int s_wlock_dispatch_loop_context(struct dispatch_loop_context *context) {
+    return aws_rw_lock_wlock(&context->lock);
+}
+
+static int s_wunlock_dispatch_loop_context(struct dispatch_loop_context *context) {
+    return aws_rw_lock_wunlock(&context->lock);
+}
+
+static int s_lock_cross_thread_data(struct dispatch_loop *loop) {
+    return aws_mutex_lock(&loop->synced_data.lock);
+}
+
+static int s_unlock_cross_thread_data(struct dispatch_loop *loop) {
+    return aws_mutex_unlock(&loop->synced_data.lock);
+}
+
+static int s_lock_service_entries(struct dispatch_loop_context *context) {
+    return aws_mutex_lock(&context->scheduling_state.services_lock);
+}
+
+static int s_unlock_service_entries(struct dispatch_loop_context *context) {
+    return aws_mutex_unlock(&context->scheduling_state.services_lock);
+}
+
+// Not sure why use 7 as the default queue size. Just follow what we used in task_scheduler.c
+static const size_t DEFAULT_QUEUE_SIZE = 7;
+static int s_compare_timestamps(const void *a, const void *b) {
+    uint64_t a_time = (*(struct scheduled_service_entry **)a)->timestamp;
+    uint64_t b_time = (*(struct scheduled_service_entry **)b)->timestamp;
+    return a_time > b_time; /* min-heap */
+}
+
+static struct scheduled_service_entry *s_scheduled_service_entry_new(
+    struct dispatch_loop_context *context,
+    uint64_t timestamp) {
+    struct scheduled_service_entry *entry =
+        aws_mem_calloc(context->allocator, 1, sizeof(struct scheduled_service_entry));
+
+    entry->allocator = context->allocator;
+    entry->timestamp = timestamp;
+    entry->dispatch_queue_context = s_acquire_dispatch_loop_context(context);
+    aws_priority_queue_node_init(&entry->priority_queue_node);
+
+    return entry;
+}
+
+/**
+ * The function should be wrapped around scheduling_status->lock
+ */
+static void s_scheduled_service_entry_destroy(
```

Comment:
This is only called as part of cleanup on `s_run_iteration`. Can we remove this function and add its execution code into s_run_iteration. 

If we keep the function, we should pull in the locking and unlocking of the scheduling_status->lock  into this function itself to keep the other end cleaner as well as update the comment to explain what this function is doing in the comment description above.